Indexing/CodeTreeInterfaces.cpp

```diff
@@ -114,66 +119,69 @@ class CodeTreeTIS::ResultIterator
     if(_finished) {
       return false;
     }
-    void* data=_matcher->next();
-    _found=static_cast<TermCodeTree::TermInfo*>(data);
+    _found = _matcher->next();
     if(!_found) {
       _finished=true;
     }
     return _found;
   }
 
-  QueryRes<ResultSubstitutionSP, TermLiteralClause> next()
+  QueryRes<ResultSubstitutionSP, Data> next()
   {
     ASS(_found);
 
     ResultSubstitutionSP subs;
     if (_retrieveSubstitutions) {
-      _resultNormalizer->reset();
-      _resultNormalizer->normalizeVariables(_found->term);
+      if constexpr (!is_indexed_data_normalized<Data>::value) {
+        _resultNormalizer->reset();
+        _resultNormalizer->normalizeVariables(_found->term);
+      }
       subs = ResultSubstitutionSP(_subst, /* nondisposable */ true);
     }
-    auto out = QueryRes<ResultSubstitutionSP, TermLiteralClause>(subs, _found);
+    auto out = QueryRes<ResultSubstitutionSP, Data>(subs, _found);
     _found=0;
     return out;
   }
 private:
 
-  CodeTreeSubstitution* _subst;
+  CodeTreeSubstitution<Data>* _subst;
   Recycled<Renaming> _resultNormalizer;
   bool _retrieveSubstitutions;
-  TermCodeTree::TermInfo* _found;
+  Data* _found;
   bool _finished;
   CodeTreeTIS* _tree;
-  Recycled<TermCodeTree::TermMatcher> _matcher;
+  Recycled<typename TermCodeTree<Data>::TermMatcher> _matcher;
 };
 
-void CodeTreeTIS::_insert(TypedTermList t, Literal* lit, Clause* cls)
+template<class Data>
+void CodeTreeTIS<Data>::handle(Data data, bool insert)
```

Comment:
Consider lifting this to the header file for tiny perf wins, I expect `insert` is often a constant at the call site.

Reply:
Forgive my ignorance, but what exactly happens when I lift this to the header? Can it be inlined?

Reply:
Exactly, other translation units have access to the definition so they can choose to inline it if they like.